File: ScaleformUI_Csharp/Menus/UIMenu/Panels/UIMenuStatisticsPanel.cs
using System.Reflection.Emit;

namespace ScaleformUI.Menu
{
    public class UIMenuStatisticsPanel : UIMenuPanel
    {
        internal readonly List<StatisticsForPanel> Items;

        public UIMenuStatisticsPanel()
        {
            Items = new List<StatisticsForPanel>();
        }

        public void AddStatistics(string Name, float val)
        {
            float _value = val;
            if (_value > 100)
                _value = 100;
            if (_value < 0)
                _value = 0;
            StatisticsForPanel item = new(Name, _value);
            Items.Add(item);
            if (ParentItem != null && ParentItem.Parent != null)
            {
                int it = ParentItem.Parent.MenuItems.IndexOf(ParentItem);
                ParentItem.Parent.SendPanelsToItemScaleform(it, true);
            }
        }

        public float GetPercentage(int ItemId)
        {
            return Items[ItemId].Value;
        }

        public void UpdateStatistic(int ItemId, float number)
        {
            Items[ItemId].Value = number;
            if (Items[ItemId].Value > 100)
                Items[ItemId].Value = 100;
            if (Items[ItemId].Value < 0)
                Items[ItemId].Value = 0;
            if (ParentItem != null && ParentItem.Parent != null)
            {
                int it = ParentItem.Parent.MenuItems.IndexOf(ParentItem);
                ParentItem.Parent.SendPanelsToItemScaleform(it, true);
            }
        }
    }

    public class StatisticsForPanel
    {
        public string Text { get; set; }
        public float Value { get; set; }

        public StatisticsForPanel(string label, float value)
        {
            Text = label;
            Value = value;
            if (Value > 100)
                Value = 100;
            if (Value < 0)
                Value = 0;
        }

        public override string ToString()
        {
            return $"{Text}:{Value}";
        }
    }
}

File: ScaleformUI_Lua/src/Menus/UIMenu/panels/UIMenuStatisticsPanel.lua
UIMenuStatisticsPanel = setmetatable({}, UIMenuStatisticsPanel)
UIMenuStatisticsPanel.__index = UIMenuStatisticsPanel
UIMenuStatisticsPanel.__call = function() return "UIMenuPanel", "UIMenuStatisticsPanel" end

---@class UIMenuStatisticsPanel
---@field public Items table
---@field public ParentItem UIMenuItem -- required
---@field public SetParentItem fun(self:UIMenuStatisticsPanel, item:UIMenuItem):UIMenuItem -- required

function UIMenuStatisticsPanel.New(items)
    local _UIMenuStatisticsPanel = {
        Items = items or {},
        ParentItem = nil, -- required
    }
    return setmetatable(_UIMenuStatisticsPanel, UIMenuStatisticsPanel)
end

---Set the parent item of the panel
---@param item UIMenuItem
---@return UIMenuItem
function UIMenuStatisticsPanel:SetParentItem(item) -- required
    if not item() ~= nil then
        self.ParentItem = item
    end
    return self.ParentItem
end

---Add a statistic to the panel
---@param name string -- The name of the statistic
---@param value number -- Must be between 0 and 100
function UIMenuStatisticsPanel:AddStatistic(name, value) -- required
    if name ~= nil and name ~= "" and value ~= nil then
        if value > 100 then
            value = 100
        elseif value < 0 then
            value = 0
        end
        table.insert(self.Items, { ['name'] = name, ['value'] = value })
        if self.ParentItem ~= nil and self.ParentItem:SetParentMenu() ~= nil and self.ParentItem:SetParentMenu():Visible() then
            local it = IndexOf(self.ParentItem:SetParentMenu().Items, self.ParentItem)
            self.ParentItem:SetParentMenu():SendPanelsToItemScaleform(it, true)
        end
    end
end

---Get the percentage of a statistic
---@param index number -- The index of the statistic 1-4
---@return number
function UIMenuStatisticsPanel:GetPercentage(index)
    if index ~= nil then
        if index > #self.Items then
            index = #self.Items
        elseif index < 1 then
            index = 1
        end
        return self.Items[index].value
    end
    return 0
end

---Update a statistic
---@param index number -- The index of the statistic starting from 1
---@param value number -- Must be a between 0 and 100
function UIMenuStatisticsPanel:UpdateStatistic(index, value)
    if value ~= nil then
        if value > 100 then
            value = 100
        elseif value < 0 then
            value = 0
        end
        self.Items[index].value = value
        if self.ParentItem ~= nil and self.ParentItem:SetParentMenu() ~= nil and self.ParentItem:SetParentMenu():Visible() then
            local it = IndexOf(self.ParentItem:SetParentMenu().Items, self.ParentItem)
            self.ParentItem:SetParentMenu():SendPanelsToItemScaleform(it, true)
        end
    end
end
